XMonad/Actions/EasyMotion.hs

```diff
@@ -0,0 +1,378 @@
+  where
+    rw = max (fi w) (fi th)
+    rh = max (fi h) (fi th)
+
+-- | Create overlay windows the minimum size to contain their key chord
+textSize :: Position -> Rectangle -> Rectangle
+textSize th r = Rectangle { rect_width  = fi th
+                          , rect_height = fi th
+                          , rect_x      = rect_x r + (fi (rect_width r) - fi th) `div` 2
+                          , rect_y      = rect_y r + (fi (rect_height r) - fi th) `div` 2 }
+
+-- | Create overlay windows the full width of the window they select, the minimum height to contain
+--   their chord, and a proportion of the distance from the top of the window they select
+bar :: RealFrac f => f -> Position -> Rectangle -> Rectangle
+bar f th r = Rectangle { rect_width  = rect_width r
+                       , rect_height = fi th
+                       , rect_x      = rect_x r
+                       , rect_y      = rect_y r + round (f' * (fi (rect_height r) - fi th)) }
+                         -- clamp f in [0,1] as other values will appear to lock up xmonad
+                         -- as the overlay will be displayed off-screen
+  where f' = min 0.0 $ max f 1.0
+
+-- | Handles overlay display and window selection. Called after config has been sanitised.
+handleSelectWindow :: EasyMotionConfig -> X (Maybe Window)
+handleSelectWindow EMConf { sKeys = AnyKeys [] } = return Nothing
+handleSelectWindow c = do
+  f <- initXMF $ font c
+  th <- (\(asc, dsc) -> asc + dsc + 2) <$> (textExtentsXMF f (concatMap keysymToString (allKeys . sKeys $ c)))
+  XConf { theRoot = rw, display = dpy } <- ask
+  XState { mapped = mappedWins, windowset = ws } <- get
+  -- build overlays depending on key configuration
+  overlays <- case sKeys c of
+                AnyKeys ks -> buildOverlays ks <$> sortedOverlayWindows
+                  where
+                    visibleWindows :: [Window]
+                    visibleWindows = toList mappedWins
+                    sortedOverlayWindows :: X [OverlayWindow]
+                    sortedOverlayWindows = sortOverlayWindows <$> buildOverlayWindows dpy th visibleWindows
+                PerScreenKeys m -> fmap concat $ sequence $ M.foldr (:) [] $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
+                  where
+                    screenById :: ScreenId -> Maybe (W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail)
+                    screenById sid = L.find ((== sid) . screen) (W.current ws : W.visible ws)
+                    visibleWindowsOnScreen :: ScreenId -> [Window]
+                    visibleWindowsOnScreen sid = L.filter (`elem` (toList mappedWins)) $ W.integrate' $ (screenById sid) >>= W.stack . W.workspace
+                    sortedOverlayWindows :: ScreenId -> X [OverlayWindow]
+                    sortedOverlayWindows sid = sortOverlayWindows <$> (buildOverlayWindows dpy th $ visibleWindowsOnScreen sid)
```

Comment:
I would indent this (just for readability) a little more like 

``` haskell
  overlays <- case sKeys c of
    AnyKeys ks -> buildOverlays ks <$> sortedOverlayWindows
     where
      visibleWindows :: [Window]
      [...]
```

Much less wasted horizontal space.  Also, Haskell people are quite fond of aligning things instead of having long lines, i.e. instead of 

``` Haskell
PerScreenKeys m -> fmap concat $ sequence $ M.foldr (:) [] $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
```

you could write

``` haskell
PerScreenKeys m ->
  fmap concat
    $ sequence
    $ M.foldr (:) []
    $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
```

This is your module, so do what you feel like fits best; we're way beyond enforcing any kind of style guide around here :)

Reply:
Regarding the first suggestion: I prefer that too- I suspect this is another autoindent casualty, except in this case I probably would've assumed my editor indent was accepted style, rather than just having made the mistake of not noticing it!

Regarding the second: much nicer, thanks.